x/gamm/keeper/internal/events/emit.go

```diff
@@ -0,0 +1,70 @@
+package events
```

Comment:
Having it in `types` allows it to be imported from any other package in the project. However, we only should be able to emit these events from the gamm keeper or message server. Therefore, making this package internal to the gamm `keeper` package allows for avoiding accidental imports anywhere else.

Another option could be is to simply define it as an "unexported" function in a separate file inside gamm keeper. With that, there are 2 issues:
- We would have to use `export_test.go` pattern to be able to export and test these event functions
- In my opinion, our keeper package is already bloated with functionality. `events` felt like a separate component that could be extracted

At the same time, I liked the ability to call these functions as `events.Emit...()` (by prefixing with `events.`. This makes it more visible than `k.emit...()` that events are being emitted. 

I'm planning to incrementally move all remaining events from the gamm keeper into this internal package and unit test them in PRs to come.

Let me know what you think

Reply:
In general, I prefer to be consistent with other modules. Typically, events are defined in `types/events.go`. 

Now I don't have a strong opinion here, but I will state that having events "public" can be a good and desirable thing. E.g. I've wrote many tools that exported and scraped on-chain data, where I used events to query for txs/blocks. Being able to directly reference the events was immensely useful instead of having to redefine them.  

So for this reason, I say just keep them in `types/`.

Reply:
agreed here!

Reply:
Thanks for the feedback.

I agree about the need for having event names public by clients/tools, For this reason, they are still kept in types as per suggestion:
https://github.com/osmosis-labs/osmosis/blob/ad633616412c32d8918411e06120b65a9fd84a07/x/gamm/types/events.go#L3-L14

This is also consistent with other modules such as `epochs`:
https://github.com/osmosis-labs/osmosis/blob/716c6e4e0cb01bc089559bc3914fa17c93f73f8b/x/epochs/types/events.go#L1-L9

Contrary to event names, clients should never be able to emit the events. This is something that only the gamm module and internal logic should be able to do. That's why this logic was moved to an internal `events` package that is never exposed to outside packages or tools.

Having it in `types` would be information leakage because we are exposing functionality that is not meant to be used by any other module, and, as a result, opening up a vector to make a mistake by accidentally emitting events from somewhere else.

@alexanderbez @mattverse @ValarDragon  please let me know what you think